aws/rust-runtime/aws-config/src/http_credential_provider.rs

```diff
@@ -176,13 +176,20 @@ impl ClassifyRetry<SdkSuccess<Credentials>, SdkError<CredentialsError>>
                 RetryKind::Error(ErrorKind::TransientError)
             }
             // non-parseable 200s
-            Err(SdkError::ServiceError {
-                err: CredentialsError::Unhandled { .. },
-                raw,
-            }) if raw.http().status().is_success() => RetryKind::Error(ErrorKind::ServerError),
+            Err(SdkError::ServiceError(context))
+                if matches!(context.err(), CredentialsError::Unhandled { .. })
```

Comment:
I didn't know where to put this comment, but the RFC says " The source implementation should not use the catch all (_) match arm, as this makes it easy to miss adding a new error variant's source at a later date.". I found a few files that still have this (just pointing them out in case you want to fix those in this review, otherwise we can just leave them for now):

```
aws-config/src/ecs.rs

impl Error for EcsConfigurationErr {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match &self {
            EcsConfigurationErr::InvalidRelativeUri { err, .. } => Some(err),
            EcsConfigurationErr::InvalidFullUri { err, .. } => Some(err),
            _ => None,
        }
    }
}
```

aws_config/src/imds/client.rs
```
impl Error for ImdsError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match &self {
            ImdsError::FailedToLoadToken(inner) => Some(inner),
            _ => None,
        }
    }
}
```

There are more but they have a lot of variants so maybe let's leave them for now e.g.:

```
impl Error for SigningStageError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match &self {
            SigningStageError::SigningFailure(err) => Some(err.as_ref()),
            _ => None,
        }
    }
}
```

```
impl Error for CredentialsError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            CredentialsError::Unhandled { cause }
            | CredentialsError::ProviderError { cause }
            | CredentialsError::InvalidConfiguration { cause } => Some(cause.as_ref() as _),
            CredentialsError::CredentialsNotLoaded { context } => Some(context.as_ref() as _),
            _ => None,
        }
    }
}
```

```
impl std::error::Error for RetryConfigErr {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        use RetryConfigErr::*;
        match self {
            InvalidRetryMode { source, .. } => Some(source),
            FailedToParseMaxAttempts { source, .. } => Some(source),
            _ => None,
        }
    }
}
```

Reply:
These are likely covered in one of the other error refactor PRs (probably the `aws-config` one), but I'll double check each.

Reply:
- `CredentialsError` is fixed in #1934
- The `EcsConfigurationErr` matcher in `source` is wrong. I'm added a comment to #1934 to resolve it there
- `ImdsError` is fixed in #1934 
- `SigningStageError` needs to be fixed. Added a comment in #1922 to resolve there
- `RetryConfigErr` is fixed in #1934